cpp/sophus/image/dyn_image.h

```diff
@@ -59,13 +59,27 @@ class DynImage : public DynImageView<TPredicate> {
 
   /// Create type-image image from provided size and pixel type.
   /// Pixel data is left uninitialized
-  DynImage(ImageSize const& size, PixelFormat const& pixel_type)
-      : DynImage(MutDynImage<TPredicate, TAllocator>(size, pixel_type)) {}
+  DynImage(ImageSize const& size, PixelFormat const& pixel_format)
+      : DynImage(MutDynImage<TPredicate, TAllocator>(size, pixel_format)) {
+    SOPHUS_ASSERT(TPredicate::isFormatValid(pixel_format));
+  }
 
   /// Create type-image image from provided layout and pixel type.
   /// Pixel data is left uninitialized
-  DynImage(ImageLayout const& layout, PixelFormat const& pixel_type)
-      : DynImage(MutDynImage<TPredicate, TAllocator>(layout, pixel_type)) {}
+  DynImage(ImageLayout const& layout, PixelFormat const& pixel_format)
+      : DynImage(MutDynImage<TPredicate, TAllocator>(layout, pixel_format)) {
+    SOPHUS_ASSERT(TPredicate::isFormatValid(pixel_format));
+  }
+
+  template <class TOtherPredicate>
+  static auto tryFrom(DynImage<TOtherPredicate, TAllocator> other_image)
+      -> Expected<DynImage<TPredicate, TAllocator>> {
+    if (!TPredicate::isFormatValid(other_image.pixelFormat())) {
+      return SOPHUS_UNEXPECTED("pixel format does not satisfy predicate");
+    }
+    return DynImage<TPredicate, TAllocator>(
+        other_image.layout(), other_image.pixelFormat(), other_image.shared_);
+  }
 
```

Comment:
From an API perspective, it's maybe a little surprising to the user that there's a mix of constructors and static member functions

Reply:
Also naming-wise I wonder if this could have a naming convention closer to `makeCopyFrom` for consistency

Reply:
Yeah, good call out. In general, I aim to use implicit(!) constructors were convenient (e.g. Image from MutImage&&, DynImage from Image), but  static factory methods anywhere else for readability.

Note, that tryFrom is not making a deep copy, just a copy over the underlying shared-ptr.